File: validators/validate-survey-html-runtime.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');

function makeReporter() {
  const errors = [];
  const warnings = [];
  return {
    error(path, message) { errors.push({ path, message }); },
    warn(path, message) { warnings.push({ path, message }); },
    result(extra = {}) { return { valid: errors.length === 0, errors, warnings, ...extra }; },
  };
}

function has(text, pattern) {
  return typeof pattern === 'string' ? text.includes(pattern) : pattern.test(text);
}

function checkPattern(text, reporter, key, message, pattern, level = 'error') {
  if (!has(text, pattern)) reporter[level](key, message);
}

function extractSurveySchemaLiteral(html) {
  const marker = 'const surveySchema =';
  const start = html.indexOf(marker);
  if (start === -1) return null;
  const from = html.indexOf('{', start);
  if (from === -1) return null;
  let depth = 0;
  let inSingle = false;
  let inDouble = false;
  let inTemplate = false;
  let escaped = false;
  for (let i = from; i < html.length; i++) {
    const ch = html[i];
    if (escaped) { escaped = false; continue; }
    if (ch === '\\') { escaped = true; continue; }
    if (!inDouble && !inTemplate && ch === '\'') { inSingle = !inSingle; continue; }
    if (!inSingle && !inTemplate && ch === '"') { inDouble = !inDouble; continue; }
    if (!inSingle && !inDouble && ch === '`') { inTemplate = !inTemplate; continue; }
    if (inSingle || inDouble || inTemplate) continue;
    if (ch === '{') depth++;
    if (ch === '}') {
      depth--;
      if (depth === 0) return html.slice(from, i + 1);
    }
  }
  return null;
}

function validateHtmlRuntime(html) {
  const reporter = makeReporter();

  checkPattern(html, reporter, 'html.doctype', 'HTML must include <!DOCTYPE html>.', /<!DOCTYPE html>/i);
  checkPattern(html, reporter, 'html.form', 'HTML must include a real <form> element.', /<form\b/i);
  checkPattern(html, reporter, 'html.schema', 'HTML should contain a surveySchema object.', 'const surveySchema =');
  checkPattern(html, reporter, 'runtime.render', 'HTML should render screens from schema.', 'render()');
  checkPattern(html, reporter, 'runtime.assemblePayload', 'HTML should define assemblePayload().', 'function assemblePayload()');
  checkPattern(html, reporter, 'runtime.validateQuestion', 'HTML should define validateQuestion().', 'function validateQuestion(');
  checkPattern(html, reporter, 'runtime.bindEvents', 'HTML should define bindEvents().', 'function bindEvents()');
  checkPattern(html, reporter, 'runtime.childVisibility', 'HTML should manage child visibility.', 'function updateChildVisibility()');
  checkPattern(html, reporter, 'runtime.localStorageSet', 'HTML should persist step cache with localStorage.setItem.', 'localStorage.setItem(');
  checkPattern(html, reporter, 'runtime.localStorageRemove', 'HTML should clear cache with localStorage.removeItem after submit.', 'localStorage.removeItem(');
  checkPattern(html, reporter, 'runtime.resumePrompt', 'HTML should support checkpoint resume prompt from localStorage cache.', 'function openResumePrompt(');
  checkPattern(html, reporter, 'runtime.consoleLog', 'HTML should output payload with console.log.', /console\.log\((payload|assemblePayload\(\))\)/);
  checkPattern(html, reporter, 'runtime.optionId', 'HTML should preserve option ids in DOM using data-option-id.', 'data-option-id');
  checkPattern(html, reporter, 'runtime.childId', 'HTML should preserve child ids in DOM using data-child-id.', 'data-child-id');
  checkPattern(html, reporter, 'runtime.exclusive', 'HTML should implement checkbox exclusive logic.', 'dataset.exclusive');
  checkPattern(html, reporter, 'runtime.mutual', 'HTML should implement checkbox mutual-exclusion logic.', 'dataset.mutualExclusion');
  checkPattern(html, reporter, 'runtime.submitIntercept', 'HTML should intercept form submit.', "form.addEventListener('submit'");
  checkPattern(html, reporter, 'runtime.changeIntercept', 'HTML should react to checkbox/input changes.', "form.addEventListener('change'");
  checkPattern(html, reporter, 'runtime.inputIntercept', 'HTML should react to input changes for cache persistence.', "form.addEventListener('input'");
  if (has(html, 'postSubmit')) {
    checkPattern(html, reporter, 'runtime.postSubmit', 'HTML with finish postSubmit should normalize and execute post-submit actions.', 'function runPostSubmitAction(');
    checkPattern(html, reporter, 'runtime.postSubmit.redirect', 'HTML with finish postSubmit should support redirect execution.', 'function navigateAfterSubmit(');
  }

  if (has(html, 'onePageOneQuestion')) {
    checkPattern(html, reporter, 'runtime.screens', 'Step mode should define screens().', 'function screens()');
    checkPattern(html, reporter, 'runtime.show', 'Step mode should define show().', 'function show(');
    checkPattern(html, reporter, 'runtime.currentQuestion', 'Step mode should define currentQuestion().', 'function currentQuestion()');
  }

  if (has(html, 'exclusive') && !has(html, '!== e.target')) {
    reporter.warn('runtime.exclusive.detail', 'Exclusive logic exists but could not find explicit exclusion of the current checkbox from clearing.');
  }

  if (has(html, 'mutual-exclusion') && !has(html, 'otherMutual')) {
    reporter.warn('runtime.mutual.detail', 'mutual-exclusion appears in schema/text but grouped clearing logic was not clearly detected.');
  }

  if (has(html, 'renderRich(') && !has(html, 'function sanitizeRichText(')) {
    reporter.warn('runtime.richtext', 'Rich-text rendering detected without explicit sanitizeRichText() hook.');
  }
  if (has(html, 'renderRich(') && !has(html, 'sanitizeRichText(')) {
    reporter.warn('runtime.richtext.sanitizer', 'Rich-text rendering detected without sanitizer hook; production use should add a whitelist sanitizer.');
  }

  const schemaLiteral = extractSurveySchemaLiteral(html);
  if (!schemaLiteral) reporter.warn('runtime.schema.extract', 'Could not extract surveySchema literal from HTML.');
  if (schemaLiteral && /randomId\(\)/.test(schemaLiteral)) {
    reporter.warn('runtime.dynamicIds', 'surveySchema contains runtime-generated ids. Production surveys should pre-freeze ids before delivery to users.');
  }
  if (schemaLiteral && !/randomId\(\)/.test(schemaLiteral)) {
    reporter.warn('runtime.staticSchemaOnly', 'Schema literal was extracted. For stronger safety, validate it separately with validate-survey-schema.js.');
  }

  const supportsRange = has(html, 'dateRange') && has(html, 'timeRange') && has(html, 'dateTimeRange');
  if (supportsRange) checkPattern(html, reporter, 'runtime.rangeObject', 'Range values should serialize to { start, end }.', /start[\s\S]*end/);

  const summary = {
    checks: {
      doctype: has(html, /<!DOCTYPE html>/i),
      form: has(html, /<form\b/i),
      schema: has(html, 'const surveySchema ='),
      assemblePayload: has(html, 'function assemblePayload()'),
      localStorage: has(html, 'localStorage.setItem(') && has(html, 'localStorage.removeItem('),
      resumePrompt: has(html, 'function openResumePrompt('),
      exclusive: has(html, 'dataset.exclusive'),
      mutualExclusion: has(html, 'dataset.mutualExclusion'),
      childVisibility: has(html, 'function updateChildVisibility()'),
      postSubmit: has(html, 'function runPostSubmitAction('),
    }
  };

  return reporter.result(summary);
}

function main() {
  const args = process.argv.slice(2);
  const json = args.includes('--json');
  const fileArg = args.find(arg => !arg.startsWith('--'));
  if (!fileArg) {
    console.error('Usage: validate-survey-html-runtime.js /absolute/path/to/file.html [--json]');
    process.exit(1);
  }
  const filePath = path.resolve(process.cwd(), fileArg);
  let html = '';
  try {
    html = fs.readFileSync(filePath, 'utf8');
  } catch (error) {
    console.error(`Failed to read HTML file: ${error.message}`);
    process.exit(1);
  }
  const report = validateHtmlRuntime(html);
  if (json) {
    console.log(JSON.stringify(report, null, 2));
  } else {
    console.log(report.valid ? '✅ HTML runtime contract check passed.' : '❌ HTML runtime contract check failed.');
    if (report.errors.length) {
      console.log('\nErrors:');
      report.errors.forEach((item, idx) => console.log(`${idx + 1}. [${item.path}] ${item.message}`));
    }
    if (report.warnings.length) {
      console.log('\nWarnings:');
      report.warnings.forEach((item, idx) => console.log(`${idx + 1}. [${item.path}] ${item.message}`));
    }
  }
  process.exit(report.valid ? 0 : 1);
}

if (require.main === module) main();

module.exports = { validateHtmlRuntime, extractSurveySchemaLiteral };

File: skills/survey-creator/validators/validate_survey_html_runtime.py
#!/usr/bin/env python3
import json, re, sys
from pathlib import Path

class Reporter:
    def __init__(self):
        self.errors = []
        self.warnings = []
    def error(self, path, message):
        self.errors.append({"path": path, "message": message})
    def warn(self, path, message):
        self.warnings.append({"path": path, "message": message})
    def result(self, extra=None):
        base = {"valid": len(self.errors) == 0, "errors": self.errors, "warnings": self.warnings}
        if extra:
            base.update(extra)
        return base


def has(text, pattern):
    return pattern.search(text) is not None if hasattr(pattern, "search") else pattern in text


def check_pattern(text, reporter, key, message, pattern, level="error"):
    if not has(text, pattern):
        getattr(reporter, level)(key, message)


def extract_survey_schema_literal(html):
    marker = "const surveySchema ="
    start = html.find(marker)
    if start == -1:
        return None
    start = html.find("{", start)
    if start == -1:
        return None
    depth = 0
    in_single = in_double = in_template = escaped = False
    for i in range(start, len(html)):
        ch = html[i]
        if escaped:
            escaped = False
            continue
        if ch == "\\":
            escaped = True
            continue
        if not in_double and not in_template and ch == "'":
            in_single = not in_single
            continue
        if not in_single and not in_template and ch == '"':
            in_double = not in_double
            continue
        if not in_single and not in_double and ch == "`":
            in_template = not in_template
            continue
        if in_single or in_double or in_template:
            continue
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                return html[start:i+1]
    return None


def validate_html_runtime(html):
    reporter = Reporter()
    check_pattern(html, reporter, "html.doctype", "HTML must include <!DOCTYPE html>.", re.compile(r"<!DOCTYPE html>", re.I))
    check_pattern(html, reporter, "html.form", "HTML must include a real <form> element.", re.compile(r"<form\b", re.I))
    check_pattern(html, reporter, "html.schema", "HTML should contain a surveySchema object.", "const surveySchema =")
    check_pattern(html, reporter, "runtime.render", "HTML should render screens from schema.", "render()")
    check_pattern(html, reporter, "runtime.assemblePayload", "HTML should define assemblePayload().", "function assemblePayload()")
    check_pattern(html, reporter, "runtime.validateQuestion", "HTML should define validateQuestion().", "function validateQuestion(")
    check_pattern(html, reporter, "runtime.bindEvents", "HTML should define bindEvents().", "function bindEvents()")
    check_pattern(html, reporter, "runtime.childVisibility", "HTML should manage child visibility.", "function updateChildVisibility()")
    check_pattern(html, reporter, "runtime.localStorageSet", "HTML should persist step cache with localStorage.setItem.", "localStorage.setItem(")
    check_pattern(html, reporter, "runtime.localStorageRemove", "HTML should clear cache with localStorage.removeItem after submit.", "localStorage.removeItem(")
    check_pattern(html, reporter, "runtime.consoleLog", "HTML should output payload with console.log.", re.compile(r"console\.log\((payload|assemblePayload\(\))\)"))
    check_pattern(html, reporter, "runtime.optionId", "HTML should preserve option ids in DOM using data-option-id.", "data-option-id")
    check_pattern(html, reporter, "runtime.childId", "HTML should preserve child ids in DOM using data-child-id.", "data-child-id")
    check_pattern(html, reporter, "runtime.exclusive", "HTML should implement checkbox exclusive logic.", "dataset.exclusive")
    check_pattern(html, reporter, "runtime.mutual", "HTML should implement checkbox mutual-exclusion logic.", "dataset.mutualExclusion")
    check_pattern(html, reporter, "runtime.submitIntercept", "HTML should intercept form submit.", "form.addEventListener('submit'")
    check_pattern(html, reporter, "runtime.changeIntercept", "HTML should react to checkbox/input changes.", "form.addEventListener('change'")
    check_pattern(html, reporter, "runtime.inputIntercept", "HTML should react to input changes for cache persistence.", "form.addEventListener('input'")

    # Explicit supported question-family runtime hooks. These checks keep the fixed renderer honest
    # as new question types are added. They are intentionally template-level checks rather than
    # business-content checks.
    check_pattern(html, reporter, "runtime.radio.collect", "HTML should collect radio answers.", "question.type === 'radio'")
    check_pattern(html, reporter, "runtime.checkbox.collect", "HTML should collect checkbox answers.", "question.type === 'checkbox'")
    check_pattern(html, reporter, "runtime.input.collect", "HTML should collect input answers.", "question.type === 'input'")
    check_pattern(html, reporter, "runtime.score.render", "HTML should define score renderer.", "function renderScoreQuestion(")
    check_pattern(html, reporter, "runtime.score.collect", "HTML should collect score answers as score rows.", "question.type === 'score'")
    check_pattern(html, reporter, "runtime.score.optionId", "Score runtime should preserve score option ids.", "data-score-option-id")
    check_pattern(html, reporter, "runtime.score.display", "Score runtime should update active score display.", "function updateScoreDisplay(")
    check_pattern(html, reporter, "runtime.nps.render", "HTML should define NPS renderer.", "function renderNpsQuestion(")
    check_pattern(html, reporter, "runtime.nps.collect", "HTML should collect NPS answer as a single object.", "question.type === 'nps'")
    check_pattern(html, reporter, "runtime.nps.optionId", "NPS runtime should preserve NPS option id.", "data-nps-option")
    check_pattern(html, reporter, "runtime.nps.descRange", "NPS runtime should resolve range-based scoreDesc values.", "function scoreDescForValue(")
    check_pattern(html, reporter, "runtime.media", "HTML should support media rendering.", "function renderMedia(")

    check_pattern(html, reporter, "runtime.logic.compute", "HTML should define computeLogicState() for conditional logic.", "function computeLogicState()")
    check_pattern(html, reporter, "runtime.logic.apply", "HTML should define applyLogicRuntime() for conditional logic.", "function applyLogicRuntime(")
    check_pattern(html, reporter, "runtime.logic.jump", "HTML should preserve jump targets in runtime logic state.", "jumpTargets")
    check_pattern(html, reporter, "runtime.logic.jumpPage", "HTML should support jump_to_page logic actions.", "action.type === 'jump_to_page'")
    check_pattern(html, reporter, "runtime.logic.optionVisibility", "HTML should toggle option visibility for logic rules.", "isOptionHidden(")
    check_pattern(html, reporter, "runtime.logic.questionVisibility", "HTML should toggle question visibility for logic rules.", "isQuestionHidden(")
    if '"Pagination"' in html or "'Pagination'" in html:
        check_pattern(html, reporter, "runtime.pagination.answerable", "Pagination support should separate answerable questions from separators.", "answerableQuestions")
        check_pattern(html, reporter, "runtime.pagination.pages", "Pagination support should build manual pages from separators.", "buildManualPagesFromSeparators(")
        check_pattern(html, reporter, "runtime.pagination.screen", "Pagination support should render page screens.", "function renderManualPage(")

    if "onePageOneQuestion" in html:
        check_pattern(html, reporter, "runtime.screens", "Step mode should define screens().", "function screens()")
        check_pattern(html, reporter, "runtime.show", "Step mode should define show().", "function show(")
        check_pattern(html, reporter, "runtime.currentQuestion", "Step mode should define currentQuestion().", "function currentQuestion()")
    if "exclusive" in html and "!== e.target" not in html:
        reporter.warn("runtime.exclusive.detail", "Exclusive logic exists but could not find explicit exclusion of the current checkbox from clearing.")
    if "mutual-exclusion" in html and "otherMutual" not in html:
        reporter.warn("runtime.mutual.detail", "mutual-exclusion appears in schema/text but grouped clearing logic was not clearly detected.")
    if "renderRich(" in html and "function sanitizeRichText(" not in html:
        reporter.warn("runtime.richtext", "Rich-text rendering detected without explicit sanitizeRichText() hook.")
    if "renderRich(" in html and "sanitizeRichText(" not in html:
        reporter.warn("runtime.richtext.sanitizer", "Rich-text rendering detected without sanitizer hook; production use should add a whitelist sanitizer.")
    schema_literal = extract_survey_schema_literal(html)
    if not schema_literal:
        reporter.warn("runtime.schema.extract", "Could not extract surveySchema literal from HTML.")
    elif "randomId()" in schema_literal:
        reporter.warn("runtime.dynamicIds", "surveySchema contains runtime-generated ids. Production surveys should pre-freeze ids before delivery to users.")
    else:
        reporter.warn("runtime.staticSchemaOnly", "Schema literal was extracted. For stronger safety, validate it separately with validate_survey_schema.py.")
    if all(x in html for x in ["dateRange", "timeRange", "dateTimeRange"]):
        check_pattern(html, reporter, "runtime.rangeObject", "Range values should serialize to { start, end }.", re.compile(r"start[\s\S]*end"))
    summary = {"checks": {
        "doctype": has(html, re.compile(r"<!DOCTYPE html>", re.I)),
        "form": has(html, re.compile(r"<form\b", re.I)),
        "schema": "const surveySchema =" in html,
        "assemblePayload": "function assemblePayload()" in html,
        "localStorage": "localStorage.setItem(" in html and "localStorage.removeItem(" in html,
        "exclusive": "dataset.exclusive" in html,
        "mutualExclusion": "dataset.mutualExclusion" in html,
        "childVisibility": "function updateChildVisibility()" in html,
        "score": "function renderScoreQuestion(" in html and "question.type === 'score'" in html,
        "nps": "function renderNpsQuestion(" in html and "question.type === 'nps'" in html,
        "media": "function renderMedia(" in html,
        "logic": "function computeLogicState()" in html and "function applyLogicRuntime(" in html,
        "pagination": "buildManualPagesFromSeparators(" in html and "answerableQuestions" in html,
    }}
    return reporter.result(summary)


def print_human(report):
    print("✅ HTML runtime contract check passed." if report["valid"] else "❌ HTML runtime contract check failed.")
    if report["errors"]:
        print("\nErrors:")
        for i, item in enumerate(report["errors"], start=1):
            print(f"{i}. [{item['path']}] {item['message']}")
    if report["warnings"]:
        print("\nWarnings:")
        for i, item in enumerate(report["warnings"], start=1):
            print(f"{i}. [{item['path']}] {item['message']}")


def main():
    args = sys.argv[1:]
    json_output = "--json" in args
    file_arg = next((a for a in args if not a.startswith("--")), None)
    if not file_arg:
        print("Usage: validate_survey_html_runtime.py /absolute/path/to/file.html [--json]", file=sys.stderr)
        sys.exit(1)
    try:
        html = Path(file_arg).read_text(encoding="utf-8")
    except FileNotFoundError as e:
        print(f"Failed to read HTML file: {e}", file=sys.stderr)
        sys.exit(1)
    report = validate_html_runtime(html)
    if json_output:
        print(json.dumps(report, ensure_ascii=False, indent=2))
    else:
        print_human(report)
    sys.exit(0 if report["valid"] else 1)


if __name__ == "__main__":
    main()
